Add README.md

davidebeatrici opened the pull request.

Krzmbrzl commented on README.md:
```diff
@@ -0,0 +1,812 @@
+
+### Endian
+
+Swaps the bit order on 16, 32 and 64 bits integers.
```
```suggestion
Swaps the byte order on 16, 32 and 64 bits integers.
```
The order of bits in individual bytes is unaffected

davidebeatrici replied:
Oops, a lapsus.

Krzmbrzl commented on README.md:
```diff
@@ -0,0 +1,812 @@
+- `NetHeader` (+ data size if UDP)
+
+Allocating extra space for the internal buffer is possible and it's what `Peer` does internally for UDP, to improve efficiency.  
+The buffer size is set to 1024 bytes, which is the formal maximum for our protocol, and the `Pack` object is recycled for every packet received.
```
We should drop this formal constraint from our protocol and thus also not use it in libmumble.

davidebeatrici replied:
No problem, we can just pass the desired buffer size when starting the UDP engine.

davidebeatrici replied:
d5215ef0a3f8162ad816d082086a57744fbc1683

Krzmbrzl replied:
Wel that's not really what I meant. With this the package size still has a hard limit depending on what the library user chose to use. Instead we should adapt the size to however large the incoming packet is (with some sanity checking to make sure maliciously crafted packages can drown the peer)

davidebeatrici replied:
In terms of efficiency it's much better to know in advance the maximum reasonable packet size.

It's fairly straightforward to calculate that, since we assume only Opus is used.

Krzmbrzl replied:
I know, but having a hard-limit programmed into the application makes future extensions of the protocol difficult as older clients will outright reject any packet that exceeds what they consider the maximum package size.

And if you reuse the buffer for the packages, I think you get about the same efficiency as you essentially only need to resize the buffer once or twice perhaps and then its size will remain constant.

davidebeatrici replied:
> I know, but having a hard-limit programmed into the application makes future extensions of the protocol difficult as older clients will outright reject any packet that exceeds what they consider the maximum package size.

Fair point. Could you open an issue for it, please?

> And if you reuse the buffer for the packages, I think you get about the same efficiency as you essentially only need to resize the buffer once or twice perhaps and then its size will remain constant.

Yes, of course.

Krzmbrzl commented on README.md:
```diff
@@ -0,0 +1,812 @@
+`endpoint` can be filled in order to use a specific source endpoint.  
+Especially useful when you have multiple Internet connections and want to force a specific one.
+
+Returns an [`std::pair`](https://en.cppreference.com/w/cpp/utility/pair) containing the result code and a socket handle (also called "file descriptor").  
```
This seems odd. We are writing C++ and thus errors should be communicated by means of exceptions and not via result codes.

davidebeatrici replied:
Honestly, I've ever only used exceptions when strictly required, either by the standard library or an external one.

The "modern" way would be to use [`std::expected`](https://en.cppreference.com/w/cpp/utility/expected), as it's officially introduced in C++23.
A potential alternative is [`Boost.Outcome`](https://www.boost.org/doc/libs/1_80_0/libs/outcome/doc/html/index.html).

Krzmbrzl replied:
Exceptions are still the best way of error handling, imo as they can't be ignored. They have to be handled or else the program terminates. It is unfortunate that C++ compilers don't warn you about potentially unhandled exceptions, though :shrug: 

Using monads is not necessarily the "modern" way of doing things. It's just another way.
However, since we don't have monads yet anyway, I would argue that exceptions are the best thing we got for the moment (certainly much better than error codes)

davidebeatrici replied:
I would argue it also depends on the way they're presented, in this library the behavior is very consistent.

On the other hand, unhandled exceptions generally trigger an automatic backtrace, which may be desiderable.

Krzmbrzl replied:
Even if the error codes are very consistently returned, it is just so easy to forget to check them. I've seen and also experienced this so many times and the result of that is a program that doesn't do error checking, which more often than not will lead to the program crashing somewhere down the line and then it will be much more difficult to figure out where the error came from. With exceptions, you get a crash right when and where the error appears (if you don't handle it).

Krzmbrzl reviewed (commented):
Just a few thoughts I had on a first glance

davidebeatrici committed:
Add "uses" prefix to the name of functions that return a boolean state

For clarity, as without the prefix another value could be expected instead.
For example, the amount of bytes added in the case of Crypt::padding().

Co-authored-by: Robert Adam <dev@robert-adam.de>

davidebeatrici committed:
Add README.md
